### PR TITLE
Generate tmp-symbols.c, pass conditionals to tmp-funcs.h

### PR DESCRIPTION
**Pull request by codebybrett**
Generated tmp-symbols.c  replaces e-func-symbols.c.

Generating tmp-symbols.c required passing through relevant conditional directives. The logic to pass through conditional directives is now also used during generation of tmp-funcs.h.

Not all conditional directives scanned are passed through. Irrelevant directives are removed to aid readability (their presence should not affect compilation).

### CODE REVIEW

**Comment by hostilefork** [comment]
This looks great!... @zsx can you review this and do the merge?

File: make/CMakeLists.txt [hunk 1]
```diff
@@ -279,7 +279,7 @@ set (CORE_SOURCE
     ${CORE_DIR}/d-print.c
     ${CORE_DIR}/d-stack.c
     ${CORE_DIR}/d-trace.c
-    ${CORE_DIR}/e-func-symbols.c
+    ${CORE_DIR}/tmp-symbols.c
     ${CORE_DIR}/f-blocks.c
```
Comment (zsx) [inline make/CMakeLists.txt]: This change is needed for this commit (correct source file directory, and add it to the list of generated files):
```
diff --git a/make/CMakeLists.txt b/make/CMakeLists.txt
index 74fabbd..366b391 100644
--- a/make/CMakeLists.txt
+++ b/make/CMakeLists.txt
@@ -279,7 +279,7 @@ set (CORE_SOURCE
     ${CORE_DIR}/d-print.c
     ${CORE_DIR}/d-stack.c
     ${CORE_DIR}/d-trace.c
-    ${CORE_DIR}/tmp-symbols.c
+    ${CORE_GENERATED_DIR}/tmp-symbols.c
     ${CORE_DIR}/f-blocks.c
     ${CORE_DIR}/f-deci.c
     ${CORE_DIR}/f-dtoa.c
@@ -421,6 +421,7 @@ set(HEADER_OUTPUT
     ${TOP_GENERATED_SRC_DIR}/include/tmp-funcs.h
     ${TOP_GENERATED_SRC_DIR}/include/tmp-funcargs.h
     ${TOP_GENERATED_SRC_DIR}/include/tmp-strings.h
+    ${CORE_GENERATED_DIR}/tmp-symbols.c
     )
```

File: src/tools/make-headers.r [hunk 2]
```diff
@@ -81,11 +82,45 @@ emit-proto: proc [proto] [
             emit-rlib ["extern " proto "; // " the-file]
         ][
             emit-out ["extern " proto "; // " the-file]
+            either "REBTYPE" = proto-parser/proto.id [
+               emit-fsymb ["    SYM_FUNC(" proto-parser/proto.arg.1 "), // " the-file]
```
Comment (zsx) [inline src/tools/make-headers.r]: This line needs to be like (manually expand the macro REBTYPE):
               emit-fsymb ["    SYM_FUNC(T_" proto-parser/proto.arg.1 "), // " the-file]

File: src/tools/make-headers.r [hunk 3]
```diff
@@ -81,11 +82,45 @@ emit-proto: proc [proto] [
...
+        find/match directive "#endif"
+        position: find/last tail buffer "#if"
+    ][
+        rewrite-if-directives position
+    ]
+]
+
 process: func [file] [
     if verbose [probe [file]]
     data: read the-file: file
```
Comment (zsx) [inline src/tools/make-headers.r]: `proto-parser/emit-directive: :emit-directive` seems to be needed here (You might want to put it right before `proto-parser/process data`, but github doesn't allow me to comment on that line), or no directives will be emit for function symbols.

**Comment by codebybrett** [comment]
Oops. Missed the func generation after I had to manually reapply some changes.  Thanks for reviewing. I've made each of the changes and will commit in a moment.

file-base/core without files in subfolders is the same list as reading the /core folder apart from p-timer.c as you mention and also, as it happens, tmp-symbols.c - so I'm swapping it over to using file-base/core.

**Comment by zsx** [comment]
Thanks for making the changes. I think there's a change missing:
```
diff --git a/make/CMakeLists.txt b/make/CMakeLists.txt
index 74fabbd..366b391 100644
--- a/make/CMakeLists.txt
+++ b/make/CMakeLists.txt
@@ -421,6 +421,7 @@ set(HEADER_OUTPUT
     ${TOP_GENERATED_SRC_DIR}/include/tmp-funcs.h
     ${TOP_GENERATED_SRC_DIR}/include/tmp-funcargs.h
     ${TOP_GENERATED_SRC_DIR}/include/tmp-strings.h
+    ${CORE_GENERATED_DIR}/tmp-symbols.c
     )
```

Without this, CMake doesn't know how to generate the file.

Another change that I'd like to make is:
```
diff --git a/src/tools/make-headers.r b/src/tools/make-headers.r
index 96f8101..c7be2ab 100644
--- a/src/tools/make-headers.r
+++ b/src/tools/make-headers.r
@@ -20,6 +20,7 @@ print "------ Building headers"
 args: parse-args system/options/args
 output-dir: fix-win32-path to file! any [args/OUTDIR %../]
 mkdir/deep output-dir/include
+mkdir/deep output-dir/core

 r3: system/version > 2.100.0
```
This ensures the folder output-dir/core exists.

**Comment by zsx** [comment]
Actually, using file-base.r introduces another issue: b-boot.c is generated, and CMAKE allows generated files to be out of source tree, thus make-header.r needs to know if the file is generated.

So I am thinking of introduce a '+' sign into the list for this indication, and here's the realization of the idea:
```
commit 3e44b9e66c5713900c2b474b1d2fe1f9e11994c8
Author: Shixin Zeng <szeng@atronixengineering.com>
Date:   Tue Nov 15 10:20:50 2016 -0500

    Mark b-boot.c as generated file in file-base.r

    Such that make-header.r knows where to find the file, as generated files
    could be out of source tree.

diff --git a/make/CMakeLists.txt b/make/CMakeLists.txt
index 366b391..188f172 100644
--- a/make/CMakeLists.txt
+++ b/make/CMakeLists.txt
@@ -429,6 +429,7 @@ set(HEADER_DEPENDS
     ${TOOLS_DIR}/form-header.r
     ${CORE_C_FILES}
     ${NATIVE_OUTPUT}
+    ${BOOT_OUTPUT}
     )

 add_custom_command(OUTPUT
diff --git a/src/tools/file-base.r b/src/tools/file-base.r
index 61668e6..abc88f6 100644
--- a/src/tools/file-base.r
+++ b/src/tools/file-base.r
@@ -15,6 +15,8 @@ REBOL [
     }
 ]

+; NOTE: In the following file list, a (+) preceding a file is indicative that
+; the file is generated.
 core: [
     ; (A)???
     a-constants.c
@@ -22,7 +24,7 @@ core: [
     a-lib.c

     ; (B)oot
-    b-boot.c
+    + b-boot.c
     b-init.c

     ; (C)ore
diff --git a/src/tools/make-headers.r b/src/tools/make-headers.r
index 14fae50..14e5a7e 100644
--- a/src/tools/make-headers.r
+++ b/src/tools/make-headers.r
@@ -198,6 +198,17 @@ emit-out {
 }

 file-base: has load %../tools/file-base.r
+
+;prefix the generated file paths with output-dir/core
+parse file-base/core [
+       any [
+               to '+ mark: (
+                       poke mark 2 join output-dir/core [%/ mark/2]
+                       remove mark ;remove '+
+               )
+       ]
+]
+
 files: map-each file file-base/core [
     either not find/match form file {../} [to file! form file][()]
 ]
diff --git a/src/tools/make-make.r b/src/tools/make-make.r
index 2286bc8..c45e909 100644
--- a/src/tools/make-make.r
+++ b/src/tools/make-make.r
@@ -287,6 +287,10 @@ unless (
        ]
 ]

+; The + sign is sued to tell the make-header.r script that the file is generated
+; we don't care it here
+remove-each item file-base/core [item = '+]
+
 ; The + sign is used to tell the make-os-ext.r script to scan a host kit file
 ; for headers (the way make-headers.r does).  But we don't care about that
 ; here in make-make.r... so remove any + signs we find before processing.
```

**Comment by codebybrett** [comment]
I've made the two changes you suggested prior to your comment about b-boot.c and the + indicator for file-base.

I don't think I'm adding value now.  If it's possible, I would rather you made the remaining changes you think are required to get this in because I do not know how to test them and I will be travelling shortly.

**Comment by zsx** [comment]
No problem. I will pick it up, and probably rebase some of the commits such that each of them will build. Thanks again for your work.

**Comment by zsx** [comment]
This has been rebased and pushed. Thanks!

**Pull request closed without merging**